File: src/emitter.rs
use std::io::{self, Write};

mod arithmetic_operation;
mod constant;
pub mod emittable;
mod numerical_value;
mod unreachable;

pub use emittable::Emittable;

use crate::ast::{Module, NumericalType, Opcode, Program, Type};
use crate::leb128::UnsignedLeb128;
use crate::opcode::ToOpcode;

const MAGIC: &[u8] = b"\0asm";
const VERSION: &[u8] = &[0x01, 0x00, 0x00, 0x00];

const SECTION_TYPE: u8 = 0x01;
const SECTION_FUNCTION: u8 = 0x03;
const SECTION_EXPORT: u8 = 0x07;
const SECTION_CODE: u8 = 0x0a;

pub struct Emitter<W> {
    /// Where this Emitter will write to
    writer: W,
}

impl<W: Write> Emitter<W> {
    /// Emit a single byte to the writer
    pub fn emit_byte(&mut self, byte: u8) -> io::Result<usize> {
        self.emit_bytes(&[byte]).map(|()| 1)
    }

    /// Emit a sequence of bytes to the writer
    pub fn emit_bytes(&mut self, bytes: &[u8]) -> io::Result<()> {
        self.writer.write_all(bytes)
    }

    /// Emits the WASM magic constant
    fn emit_magic(&mut self) -> io::Result<()> {
        self.emit_bytes(MAGIC)
    }

    /// Emits the WASM version tag
    fn emit_version(&mut self) -> io::Result<()> {
        self.emit_bytes(VERSION)
    }

    /// Builds a new emitter with the given writer
    pub fn new(writer: W) -> Self {
        Self { writer }
    }

    /// Emit the given program to WASM
    pub fn emit_program(&mut self, program: Program) -> io::Result<()> {
        self.emit_magic()?;
        self.emit_version()?;

        for module in program.modules {
            self.emit_module(module)?;
        }

        Ok(())
    }

    fn emit_module(&mut self, module: Module) -> io::Result<()> {
        if module.functions.is_empty() {
            return Ok(());
        }

        self.emit_type_section(&module)?;
        self.emit_function_section(&module)?;
        self.emit_export_section(&module)?;
        self.emit_code_section(&module)?;

        Ok(())
    }

    fn emit_type_section(&mut self, module: &Module) -> io::Result<()> {
        let mut section_data = Vec::new();
        let mut section_emitter = Emitter::new(&mut section_data);

        section_emitter.emit_element(UnsignedLeb128::from(module.functions.len() as u64))?;

        for function in &module.functions {
            section_emitter.emit_byte(0x60)?;
            section_emitter.emit_element(UnsignedLeb128::from(function.parameters.len() as u64))?;

            for param in &function.parameters {
                section_emitter.emit_byte(type_to_byte(&param.type_))?;
            }

            if let Some(ref return_type) = function.return_type {
                section_emitter.emit_byte(0x01)?;
                section_emitter.emit_byte(type_to_byte(return_type))?;
            } else {
                section_emitter.emit_byte(0x00)?;
            }
        }

        self.emit_section(SECTION_TYPE, &section_data)
    }

    fn emit_function_section(&mut self, module: &Module) -> io::Result<()> {
        let mut section_data = Vec::new();
        let mut section_emitter = Emitter::new(&mut section_data);

        section_emitter.emit_element(UnsignedLeb128::from(module.functions.len() as u64))?;

        for (idx, _) in module.functions.iter().enumerate() {
            section_emitter.emit_element(UnsignedLeb128::from(idx as u64))?;
        }

        self.emit_section(SECTION_FUNCTION, &section_data)
    }

    fn emit_export_section(&mut self, module: &Module) -> io::Result<()> {
        let exports: Vec<_> = module
            .functions
            .iter()
            .enumerate()
            .flat_map(|(idx, func)| func.exports.iter().map(move |name| (idx, name)))
            .collect();

        if exports.is_empty() {
            return Ok(());
        }

        let mut section_data = Vec::new();
        let mut section_emitter = Emitter::new(&mut section_data);

        section_emitter.emit_element(UnsignedLeb128::from(exports.len() as u64))?;

        for (func_idx, export_name) in exports {
            section_emitter.emit_element(UnsignedLeb128::from(export_name.len() as u64))?;
            section_emitter.emit_bytes(export_name.as_bytes())?;
            section_emitter.emit_byte(0x00)?;
            section_emitter.emit_element(UnsignedLeb128::from(func_idx as u64))?;
        }

        self.emit_section(SECTION_EXPORT, &section_data)
    }

    fn emit_code_section(&mut self, module: &Module) -> io::Result<()> {
        let mut section_data = Vec::new();
        let mut section_emitter = Emitter::new(&mut section_data);

        section_emitter.emit_element(UnsignedLeb128::from(module.functions.len() as u64))?;

        for function in &module.functions {
            let mut func_data = Vec::new();
            let mut func_emitter = Emitter::new(&mut func_data);

            func_emitter
                .emit_element(UnsignedLeb128::from(function.local_variables.len() as u64))?;

            for local in &function.local_variables {
                func_emitter.emit_byte(0x01)?;
                func_emitter.emit_byte(type_to_byte(&local.type_))?;
            }

            for instruction in &function.body {
                func_emitter.emit_instruction(instruction)?;
            }

            func_emitter.emit_byte(0x0b)?;

            section_emitter.emit_element(UnsignedLeb128::from(func_data.len() as u64))?;
            section_emitter.emit_bytes(&func_data)?;
        }

        self.emit_section(SECTION_CODE, &section_data)
    }

    fn emit_instruction(&mut self, instruction: &crate::ast::Instruction) -> io::Result<()> {
        let opcode_byte = instruction.opcode.to_opcode();
        self.emit_byte(opcode_byte)?;

        match &instruction.opcode {
            Opcode::Constant(constant) => {
                self.emit_element(constant.value)?;
            }
            Opcode::Call(index) => {
                if let crate::ast::Index::Numerical(idx) = index {
                    self.emit_element(UnsignedLeb128::from(*idx as u64))?;
                }
            }
            Opcode::VariableInstruction(var_op) => {
                if let crate::ast::Index::Numerical(idx) = &var_op.index {
                    self.emit_element(UnsignedLeb128::from(*idx as u64))?;
                }
            }
            _ => {}
        }

        for arg in &instruction.arguments {
            self.emit_instruction(arg)?;
        }

        Ok(())
    }

    fn emit_section(&mut self, section_id: u8, data: &[u8]) -> io::Result<()> {
        self.emit_byte(section_id)?;
        self.emit_element(UnsignedLeb128::from(data.len() as u64))?;
        self.emit_bytes(data)?;
        Ok(())
    }

    #[cfg(test)]
    pub fn into_inner(self) -> W {
        self.writer
    }
}

fn type_to_byte(type_: &Type) -> u8 {
    match type_ {
        Type::Numerical(NumericalType::Int32) => 0x7f,
        Type::Numerical(NumericalType::Int64) => 0x7e,
        Type::Numerical(NumericalType::Float32) => 0x7d,
        Type::Numerical(NumericalType::Float64) => 0x7c,
    }
}

impl<W> Emitter<std::io::Cursor<W>> {
    #[cfg(test)]
    pub fn new_cursored(writer: W) -> Self {
        use std::io::Cursor;

        Self {
            writer: Cursor::new(writer),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::MAGIC;

    #[test]
    fn assert_correct_magic() {
        assert_eq!(MAGIC, &[0x00, 0x61, 0x73, 0x6d])
    }
}

File: src/emitter.zig
const std = @import("std");
const ast = @import("ast.zig");
const opcode = @import("opcode.zig");
const leb128 = @import("leb128.zig");

const MAGIC: []const u8 = "\x00asm";
const VERSION: []const u8 = "1000";

pub fn Emitter(comptime WriterType: type) type {
    return struct {
        const Self = @This();

        /// Where this Emitter will write to
        writer: WriterType,

        pub fn init(writer: WriterType) Self {
            return Self{ .writer = writer };
        }

        /// Emit a single byte to the writer
        pub fn emitByte(self: *Self, byte: u8) !usize {
            try self.writer.writeByte(byte);
            return 1;
        }

        /// Emit a sequence of bytes to the writer
        pub fn emitBytes(self: *Self, bytes: []const u8) !void {
            try self.writer.writeAll(bytes);
        }

        /// Emits the WASM magic constant
        fn emitMagic(self: *Self) !void {
            try self.emitBytes(MAGIC);
        }

        /// Emits the WASM version tag
        fn emitVersion(self: *Self) !void {
            try self.emitBytes(VERSION);
        }

        /// Emit the given program to WASM
        pub fn emitProgram(self: *Self, program: ast.Program) !void {
            _ = program;
            try self.emitMagic();
            try self.emitVersion();
        }

        // Emittable implementations

        /// Emit an Unreachable instruction
        pub fn emitUnreachable(self: *Self, unreachable_inst: ast.Unreachable) !usize {
            _ = unreachable_inst;
            const op = opcode.toOpcode(ast.Unreachable{});
            return try self.emitByte(op);
        }

        /// Emit a NumericalValue
        pub fn emitNumericalValue(self: *Self, value: ast.NumericalValue) !usize {
            return switch (value) {
                .int32 => |int32| try leb128.emitSignedLeb128(self, leb128.SignedLeb128.init(@as(i64, int32))),
                .int64 => |int64| try leb128.emitSignedLeb128(self, leb128.SignedLeb128.init(int64)),
                .float32 => |f32_val| {
                    const bytes = f32ToBytes(f32_val);
                    try self.emitBytes(&bytes);
                    return 4;
                },
                .float64 => |f64_val| {
                    const bytes = f64ToBytes(f64_val);
                    try self.emitBytes(&bytes);
                    return 8;
                },
            };
        }

        /// Emit a Constant
        pub fn emitConstant(self: *Self, constant: ast.Constant) !usize {
            const op = opcode.toOpcode(constant.value);

            // Emit the `const` opcode for the given value
            _ = try self.emitByte(op);

            // .. and then the actual literal
            return try self.emitNumericalValue(constant.value);
        }

        /// Emit an ArithmeticOperation
        pub fn emitArithmeticOperation(self: *Self, arith_op: ast.ArithmeticOperation) !usize {
            const op = opcode.toOpcode(arith_op);
            return try self.emitByte(op);
        }
    };
}

/// According to the WebAssembly spec, floating-point values are encoded
/// by their IEEE 754-2019 (Section 3.4) bit pattern in little endian byte order.
fn f32ToBytes(n: f32) [4]u8 {
    return @bitCast(std.mem.nativeToLittle(u32, @bitCast(n)));
}

fn f64ToBytes(n: f64) [8]u8 {
    return @bitCast(std.mem.nativeToLittle(u64, @bitCast(n)));
}

test "assert correct magic" {
    try std.testing.expectEqualSlices(u8, &[_]u8{ 0x00, 0x61, 0x73, 0x6d }, MAGIC);
}

test "emits unreachable opcode" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();

    var emitter = Emitter(@TypeOf(buffer.writer())).init(buffer.writer());
    _ = try emitter.emitUnreachable(ast.Unreachable{});

    try std.testing.expectEqualSlices(u8, &[_]u8{0x00}, buffer.items);
}

test "emits i32 const correctly" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();

    var emitter = Emitter(@TypeOf(buffer.writer())).init(buffer.writer());

    const constant = ast.Constant{
        .value = ast.NumericalValue{ .int32 = 128 },
    };

    _ = try emitter.emitConstant(constant);

    try std.testing.expectEqualSlices(
        u8,
        &[_]u8{
            // `i32.const`'s opcode
            0x41,
            // and the LEB128 for 128
            128,
            1,
        },
        buffer.items,
    );
}

test "emits i64 const correctly" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();

    var emitter = Emitter(@TypeOf(buffer.writer())).init(buffer.writer());

    const constant = ast.Constant{
        .value = ast.NumericalValue{ .int64 = 505 },
    };

    _ = try emitter.emitConstant(constant);

    try std.testing.expectEqualSlices(
        u8,
        &[_]u8{
            // `i64.const`'s opcode
            0x42,
            // and the LEB128 for 505
            249,
            3,
        },
        buffer.items,
    );
}

test "emits f32 const correctly" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();

    var emitter = Emitter(@TypeOf(buffer.writer())).init(buffer.writer());

    const constant = ast.Constant{
        .value = ast.NumericalValue{ .float32 = 5.0 },
    };

    _ = try emitter.emitConstant(constant);

    try std.testing.expectEqualSlices(
        u8,
        &[_]u8{
            // `f32.const`'s opcode
            0x43,
            // and then the LE bit pattern for 5.0
            0x00, 0x00, 0xa0, 0x40,
        },
        buffer.items,
    );
}

test "emits f64 const correctly" {
    var buffer = std.ArrayList(u8).init(std.testing.allocator);
    defer buffer.deinit();

    var emitter = Emitter(@TypeOf(buffer.writer())).init(buffer.writer());

    const constant = ast.Constant{
        .value = ast.NumericalValue{ .float64 = 25.50 },
    };

    _ = try emitter.emitConstant(constant);

    try std.testing.expectEqualSlices(
        u8,
        &[_]u8{
            // `f64.const`'s opcode
            0x44,
            // and then the LE bit pattern for 25.50
            0x00, 0x00, 0x00, 0x00, 0x00, 0x80, 0x39, 0x40,
        },
        buffer.items,
    );
}
